### PR TITLE
Extra dashboard validation for deserialised resource service messages

### PR DESCRIPTION
**Pull request by drewnoakes**
With this change, the dashboard performs extra validation on the messages it receives. This will make the source of invalid data more obvious, assisting debugging of issues.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2244)

### CODE REVIEW

**Review by JamesNK — APPROVED** [review]

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs [hunk 1]
```diff
@@ -46,28 +46,56 @@ public static string GetResourceName(ResourceViewModel resource, IEnumerable<Res
...
 
-public sealed class EnvironmentVariableViewModel(string name, string? value, bool fromSpec)
+public sealed class EnvironmentVariableViewModel
 {
-    public string Name { get; } = name;
-    public string? Value { get; } = value;
-    public bool FromSpec { get; } = fromSpec;
+    public string Name { get; }
+    public string? Value { get; }
+    public bool FromSpec { get; }
```
Comment (radical) [inline src/Aspire.Dashboard/Model/ResourceViewModel.cs:53]:
```suggestion
    public string Name { get; init; }
    public string? Value { get; init; }
    public bool FromSpec { get; init; }
```

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs [hunk 2]
```diff
@@ -46,28 +46,56 @@ public static string GetResourceName(ResourceViewModel resource, IEnumerable<Res
...
+    }
 }
 
-public sealed class EndpointViewModel(string endpointUrl, string proxyUrl)
+public sealed class EndpointViewModel
 {
-    public string EndpointUrl { get; } = endpointUrl;
-    public string ProxyUrl { get; } = proxyUrl;
+    public string EndpointUrl { get; }
+    public string ProxyUrl { get; }
```
Comment (radical) [inline src/Aspire.Dashboard/Model/ResourceViewModel.cs:70]:
```suggestion
    public string EndpointUrl { get; init; }
    public string ProxyUrl { get; init; }
```

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs [hunk 3]
```diff
@@ -46,28 +46,56 @@ public static string GetResourceName(ResourceViewModel resource, IEnumerable<Res
...
+public sealed class ResourceServiceViewModel
 {
-    public string Name { get; } = name;
-    public string? AllocatedAddress { get; } = allocatedAddress;
-    public int? AllocatedPort { get; } = allocatedPort;
+    public string Name { get; }
+    public string? AllocatedAddress { get; }
+    public int? AllocatedPort { get; }
+
+    public string AddressAndPort { get; }
```
Comment (radical) [inline src/Aspire.Dashboard/Model/ResourceViewModel.cs:88]:
```suggestion
    public string Name { get; init; }
    public string? AllocatedAddress { get; init; }
    public int? AllocatedPort { get; init; }

    public string AddressAndPort { get; init; }
```

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/Model/ResourceViewModel.cs:53]: If we make these `init` properties then the validation could be bypassed.

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/Model/ResourceViewModel.cs:70]: If we make these `init` properties then the validation could be bypassed.

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/Model/ResourceViewModel.cs:88]: If we make these `init` properties then the validation could be bypassed.

File: src/Aspire.Dashboard/Protos/Partials.cs [hunk 4]
```diff
@@ -49,5 +50,15 @@ ImmutableArray<EndpointViewModel> GetEndpoints()
...
         }
+
+        static T ValidateNotNull<T>(T value, [CallerArgumentExpression(nameof(value))] string? expression = null) where T : class
+        {
+            if (value is null)
+            {
+                throw new InvalidOperationException($"Message field '{expression}' cannot be null.");
+            }
+
+            return value;
```
Comment (radical) [inline src/Aspire.Dashboard/Protos/Partials.cs:61]: super nit:
```suggestion
            return value is null
                    ? throw new InvalidOperationException($"Message field '{expression}' cannot be null.")
                    : value;
```

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs (hunk 2 above)
Comment (radical) [reply src/Aspire.Dashboard/Model/ResourceViewModel.cs:70]: Oh, how would it get bypassed?

File: src/Aspire.Dashboard/Protos/Partials.cs (hunk 4 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/Protos/Partials.cs:61]: Could also be `return value ?? throw new ...` but I think I prefer the `if` style thanks.

**Pull request merged**